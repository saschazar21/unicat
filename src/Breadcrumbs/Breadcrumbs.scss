@import '../__styles__/variables';

.wrapper {
  font-size: fontSize(small);
}
