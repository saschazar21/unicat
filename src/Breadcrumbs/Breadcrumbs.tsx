import React, { ReactNode, createElement } from 'react';
import classnames from 'classnames';
import { ChevronIcon } from '@saschazar/unicat-icons';

import List from '../List';
import KeyGenerator from '../__tools__/key-generator';

import styles from './Breadcrumbs.scss';

export interface BreadcrumbsItems {
  active?: boolean;
  href?: string;
  name: string;
}

export interface BreadcrumbsProps {
  /** a custom className */
  className?: string;
  /** the breadcrumbs segments */
  items: BreadcrumbsItems[];
  /** a custom prefix */
  prefix?: ReactNode;
  /** display only first and last */
  small?: boolean;
}

export default (props: BreadcrumbsProps): JSX.Element => {
  const {
    className: customClassName,
    prefix: customPrefix,
    items,
    small,
  } = props;

  const prefix = customPrefix || <ChevronIcon />;
  const keygen = new KeyGenerator('Breadcrumbs');
  const className = classnames(
    { [styles.small]: small },
    styles.wrapper,
    customClassName
  );

  const renderItems = (): JSX.Element[] => {
    const breadcrumbs: BreadcrumbsItems[] =
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      small && items.length > 2 ? [items.shift()!, items.pop()!] : items;

    return breadcrumbs.map(({ active, href, name }) =>
      createElement(
        active && href ? 'a' : 'span',
        Object.assign({}, { key: keygen.next() }, active ? { href } : null),
        name
      )
    );
  };

  return (
    <nav>
      <List
        horizontal
        inline
        className={className}
        prefix={prefix}
        items={renderItems()}
      />
    </nav>
  );
};
