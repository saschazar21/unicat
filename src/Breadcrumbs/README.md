Renders a simple breadcrumbs list. Based on the array in the given `items` prop, it conditionally renders either a `<a />` or a `<span />`.

```jsx
const items = [
  {
    name: 'Home',
  },
  {
    href: '#components',
    name: 'Components',
  },
  {
    active: true,
    href: '#button',
    name: 'Button',
  },
];

<Breadcrumbs items={items} />;
```
