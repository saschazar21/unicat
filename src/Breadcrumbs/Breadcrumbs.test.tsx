import React from 'react';
import { mount, shallow } from 'enzyme';

import Breadcrumbs from './Breadcrumbs';
import toJson from 'enzyme-to-json';

const items = [
  {
    name: 'Home',
  },
  {
    href: '#components',
    name: 'Components',
  },
  {
    active: true,
    href: '#button',
    name: 'Button',
  },
];

describe('Breadcrumbs', () => {
  it('renders', () => {
    const wrapper = shallow(<Breadcrumbs items={items} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('mounts a default variant', () => {
    const wrapper = mount(<Breadcrumbs items={items} />);
    expect(wrapper.find('span')).toHaveLength(2);
    expect(wrapper.find('a')).toHaveLength(1);
  });

  it('mounts a small variant', () => {
    const wrapper = mount(<Breadcrumbs small items={items} />);
    expect(wrapper.find('span')).toHaveLength(1);
    expect(wrapper.find('a')).toHaveLength(1);
  });
});
