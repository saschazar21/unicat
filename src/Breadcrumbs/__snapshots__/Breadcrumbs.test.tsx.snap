// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs renders 1`] = `
<nav>
  <List
    className="wrapper"
    distribute="normal"
    horizontal={true}
    inline={true}
    items={
      Array [
        <span>
          Home
        </span>,
        <span>
          Components
        </span>,
        <a
          href="#button"
        >
          Button
        </a>,
      ]
    }
    prefix={<ChevronIcon />}
  />
</nav>
`;
